src/main/scala/org/constellation/p2p/Download.scala

```diff
@@ -14,69 +18,177 @@ import scala.concurrent.duration._
+  }
 
-    val preExistingSnapshots = dao.snapshotPath.list.toSeq.map { _.name }
+  /** **/
+
+  private def setNodeState(nodeState: NodeState): IO[Unit] =
+    IO(dao.nodeState = nodeState)
+      .flatMap(updateMetricAndPass("nodeState", nodeState.toString))
+      .flatMap(_ => IO(PeerManager.broadcastNodeState()))
+
+  private def requestForFaucet: IO[Iterable[Response[String]]] =
+    Future
+      .sequence(
+        dao.peerInfo
+          .map(_._2.client)
+          .map(_.post("faucet", SendToAddress(dao.selfAddressStr, 500L)))
+      )
+      .toIO
+
+  private def downloadSnapshotInfo(implicit snapshotClient: APIClient): IO[SnapshotInfo] =
+    IO { logger.info(s"Downloading from: ${snapshotClient.hostName}:${snapshotClient.apiPort}") }
+      .flatMap { _ =>
+        snapshotClient
+          .getNonBlockingBytesKryo[SnapshotInfo]("info", timeout = 15.seconds)
+          .toIO
+      }
+      .flatMap { a =>
+        updateMetric("downloadExpectedNumSnapshotsIncludingPreExisting",
+          a.snapshotHashes.size.toString)
+          .map(_ => a)
+      }
 
+  private def getSnapshotHashes(snapshotInfo: SnapshotInfo): IO[Seq[String]] = {
+    val preExistingSnapshots = dao.snapshotPath.list.toSeq.map(_.name)
     val snapshotHashes = snapshotInfo.snapshotHashes.filterNot(preExistingSnapshots.contains)
 
-    dao.metrics.updateMetric("downloadExpectedNumSnapshots", snapshotHashes.size.toString)
-
-    val groupSize = (snapshotHashes.size / peerData.size) + 1
-
-    dao.metrics.updateMetric("downloadGroupSize", groupSize.toString)
+    IO.pure(snapshotHashes)
+      .flatMap(updateMetricAndPass("downloadExpectedNumSnapshots", snapshotHashes.size.toString))
+  }
 
+  private def getGroupedHashes(
+    snapshotHashes: Seq[String]
+  )(implicit peers: Peers): IO[Seq[(Seq[String], PeerData)]] = {
+    val groupSize = (snapshotHashes.size / peers.size) + 1
     val groupedHashes = snapshotHashes.grouped(groupSize).toSeq
+    val grouped = groupedHashes.zip(peers.values)
 
-    dao.metrics.updateMetric("downloadGroupedHashesSize", groupedHashes.size.toString)
+    updateMetric("downloadGroupSize", groupSize.toString)
+      .flatMap(_ => updateMetric("downloadGroupedHashesSize", groupedHashes.size.toString))
+      .flatMap(_ => updateMetric("downloadGroupedZipSize", grouped.size.toString))
+      .flatMap(_ => updateMetric("downloadGroupCheckSize", grouped.flatMap(_._1).size.toString))
+      .map(_ => grouped)
+  }
 
-    val grouped = groupedHashes.zip(peerData.values)
-    dao.metrics.updateMetric("downloadGroupedZipSize", grouped.size.toString)
+  private def processSnapshots(grouped: Seq[(Seq[String], PeerData)])(implicit peers: Peers) = IO {
+    grouped.par.map {
+      case (hashes, peer) =>
+        hashes.par.map { hash =>
+          processSnapshotHash(peers, peer.client, hash)
+        }
+    }
+  }
 
-    dao.metrics.updateMetric("downloadGroupCheckSize", grouped.flatMap(_._1).size.toString)
+  // TODO mwadon
```

Comment:
may want to add notes on this TODO -- unless this is what you're going to change next PR.

Reply:
that's the thing I'm working on right now :) parallel downloading and refactoring the part in the downloading process related to that (processing snapshots as well)